# NanJ_Carles_et_al_3aQTL_project
## 1. De novo pA site identification

   Code for average 3' UTR length calculation and pA site identification. 
   
## 2. PDUI matrix process

   Code for PDUI (distal pA site usage index) calculation, PDUI matrix quality control and imputation. 

## 3. 3'aQTL calling

   Code for PDUI normalization, confounding factor (PEER factor) detection and 3'aQTL calling.

## 4. Colocalization and MR analysis

  Code for colocalization and Mendelian Randomization analysis between different mQTLs or mQTL and disease GWAS.

## 5. downstream analysis

**5.1 APA landscape analysis:** Code for cell type-specific APA detection, functional and motif enrichment analysis of cell type-specific APAs, and _trans_ APA regulator identification.

**5.2 DAPA analysis:** Code for AD-differential APA and AD-differential-expressed gene identification.

**5.3 3aQTL features:** Code for genomic enrichment analysis of 3'aQTL leadSNP, comparison of 3' UTR length and distance of proximal-distal pA sites between g-APA and non-g-APA, comparison of 3'aQTL between single cell level (this study) and bulk level (GTEx) and comparison of g-APAs across cell types.

**5.4 3'aQTL regulator identification:** Code for enrichment analysis of ChIP-seq and eCLIP-seq binding sites in 3'aQTLs, and regression analysis between expression of predicted regulators and PDUI of the corresponding targets.

**5.5 _APOC2_ analysis:** Code for conditional analysis of AD GWAS at _APOC2_ locus.

**5.6 _SNCA_ analysis:** Code for pA site of _SNCA_ identification, its 3'aQTL calling and colocalization analysis between 3'aQTL and PD GWAS. 
